File: src/server/ecs/systems/replication.lua
local Players = game:GetService("Players")
local ReplicatedStorage = game:GetService("ReplicatedStorage")

local components = require(ReplicatedStorage.paradoxical.ecs.components)
local routes = require(ReplicatedStorage.paradoxical.ecs.routes)
local useEvent = require(ReplicatedStorage.Packages.matter).useEvent

local REPLICATED_COMPONENTS = {
	"Model",
	"Health",
}

local replicatedComponents = {}

for _, name in REPLICATED_COMPONENTS do
	replicatedComponents[components[name]] = true
end

--[=[
    @function replication
    @within SystemServer

    Server replication to client.

    Replicated components:
    - [Model]
    - [Health]
]=]
local function replication(world)
	for _, player in useEvent(Players, "PlayerAdded") do
		local payload = {}

		for entityId, entityData in world do
			local entityPayload = {}
			payload[tostring(entityId)] = entityPayload

			for component, componentData in entityData do
				if replicatedComponents[component] then
					entityPayload[tostring(component)] = { data = componentData }
				end
			end
		end

		print("Sending initial payload to", player)
		routes.MatterReplication:send(payload):to(player)
	end

	local changes = {}

	for component in replicatedComponents do
		for entityId, record in world:queryChanged(component) do
			local key = tostring(entityId)
			local name = tostring(component)

			if changes[key] == nil then
				changes[key] = {}
			end

			if world:contains(entityId) then
				changes[key][name] = { data = record.new }
			end
		end
	end

	if next(changes) then
		routes.MatterReplication:send(changes)
	end
end

return {
	system = replication,
	priority = math.huge,
}

File: src/server/systems/replication.ts
import { useEvent, World } from "@rbxts/matter";
import { Players } from "@rbxts/services";
import { matterReplication } from "shared/net";
import * as Components from "shared/ecs/components";

type ComponentName = keyof typeof Components;
type ComponentConstructor = (typeof Components)[ComponentName];

const REPLICATED_COMPONENT_NAMES: readonly ComponentName[] = ["Model", "Crouching"];

const replicatedComponents: ReadonlySet<ComponentConstructor> = REPLICATED_COMPONENT_NAMES.reduce(
	(set: Set<ComponentConstructor>, name: ComponentName) => {
		return set.add(Components[name]);
	},
	new Set(),
);

function replication(world: World): void {
	let payload: Map<string, Map<ComponentName, { data?: Components.GameComponent }>> | undefined;
	for (const [_, player] of useEvent(Players, "PlayerAdded")) {
		if (!payload) {
			payload = new Map();

			for (const [id, entityData] of world) {
				const entityPayload: Map<ComponentName, { data?: Components.GameComponent }> = new Map();
				payload.set(tostring(id), entityPayload);

				for (const [component, componentData] of entityData) {
					if (replicatedComponents.has(component)) {
						// Here we are certain that the component has the name of one of our
						// components because it exists in our set of components.
						entityPayload.set(tostring(component) as ComponentName, { data: componentData });
					}
				}
			}
		}

		print("Sending initial payload to", player);
		matterReplication.replication.sendTo(
			payload as Map<
				string,
				Map<
					string,
					{
						data: Map<string, unknown> | undefined;
					}
				>
			>,
			player,
		);
	}

	const changes: Map<string, Map<ComponentName, { data?: Components.GameComponent }>> = new Map();
	for (const component of replicatedComponents) {
		// Here we are certain that the component has the name of one of our
		// components since it came from our set.
		const name = tostring(component) as ComponentName;

		for (const [id, record] of world.queryChanged(component)) {
			const key = tostring(id);

			if (!changes.has(key)) {
				changes.set(key, new Map());
			}

			if (world.contains(id)) {
				changes.get(key)?.set(name, { data: record.new });
			}
		}
	}

	if (!changes.isEmpty()) {
		matterReplication.replication.sendToAll(
			changes as Map<
				string,
				Map<
					string,
					{
						data: Map<string, unknown> | undefined;
					}
				>
			>,
		);
	}
}

/**
 * A system that replicates all replicated components to the client.
 *
 * This system runs after after all other systems to ensure any changes made are
 * included.
 */
export = {
	system: replication,
	priority: math.huge,
};
